File: .API/IdUtil.cs
// Decompiled with JetBrains decompiler
// Type: CloudX.Shared.IdUtil
// Assembly: CloudX.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 2635624C-5F24-4EFB-ACD1-7E9C1349B2F5
// Assembly location: F:\SteamLibrary\steamapps\common\NeosVR\HeadlessClient\CloudX.Shared.dll

using BaseX;
using System;
using System.Text;

namespace CloudX.Shared
{
  public static class IdUtil
  {
    public const int MAX_NAME_LENGTH = 20;

    public static OwnerType GetOwnerType(string id)
    {
      if (id == null)
        return OwnerType.INVALID;
      if (id.StartsWith("M-"))
        return OwnerType.Machine;
      if (id.StartsWith("U-"))
        return OwnerType.User;
      return id.StartsWith("G-") ? OwnerType.Group : OwnerType.INVALID;
    }

    public static string GenerateId(OwnerType ownerType, string name = null, int randomAppend = 0)
    {
      name = name != null ? name.RemoveDiacritics().RemoveNonASCII() : (string) null;
      StringBuilder stringBuilder = new StringBuilder();
      if (name != null)
      {
        foreach (char c in name)
        {
          if (char.IsLetterOrDigit(c))
            stringBuilder.Append(c);
          if (char.IsWhiteSpace(c) || c == '_')
            stringBuilder.Append("-");
          if (stringBuilder.Length == 20)
            break;
        }
      }
      if (stringBuilder.Length == 0 || randomAppend > 0)
      {
        if (stringBuilder.Length > 0)
          stringBuilder.Append("-");
        string str = Guid.NewGuid().ToString();
        if (randomAppend > 0)
          str = str.Substring(0, randomAppend);
        stringBuilder.Append(str);
      }
      switch (ownerType)
      {
        case OwnerType.Machine:
          stringBuilder.Insert(0, "M-");
          break;
        case OwnerType.User:
          stringBuilder.Insert(0, "U-");
          break;
        case OwnerType.Group:
          stringBuilder.Insert(0, "G-");
          break;
        default:
          throw new Exception("Invalid owner type");
      }
      return stringBuilder.ToString();
    }
  }
}

File: API/IdUtil.js
const { StringBuilder } = require("./StringBuilder");
const { Char } = require("./Char");
const { OwnerType } = require("./OwnerType");
const { v4: uuidv4 } = require("uuid");
/**
 *
 * @static
 * @class IdUtil
 */
class IdUtil {
	static get MAX_NAME_LENGTH() {
		return 20;
	}
	/**
   *
   * @static
   * @param {string} id
   * @returns {OwnerType}
   * @memberof IdUtil
   */
	static GetOwnerType(id) {
		if (id == null) return OwnerType.INVALID;
		if (id.startsWith("M-")) return OwnerType.Machine;
		if (id.startsWith("U-")) return OwnerType.User;
		if (id.startsWith("G-")) return OwnerType.Group;
		return OwnerType.INVALID;
	}
	/**
   *
   * @static
   * @param {OwnerType} ownerType
   * @param {string} [name=null]
   * @param {number} [randomAppend=0]
   * @memberof IdUtil
   */
	static GenerateId(ownerType, name = null, randomAppend = 0) {
		name =
      name != null
      	? name
      		.normalize("NFD")
      		.replace(/[\u0300-\u036f]/g, "")
      		.replace(/[\u{0080}-\u{FFFF}]/gu, "")
      	: null;
		var stringBuilder = new StringBuilder();
		if (name != null) {
			for (/** @type string */ let c of name) {
				if (Char.IsLetterOrDigit(c)) stringBuilder.Append(c);
				if (Char.IsWhiteSpace(c) || c === "_") stringBuilder.Append("-");
				if (stringBuilder.Length === 20) break;
			}
		}
		if (stringBuilder.Length === 0 || randomAppend > 0) {
			if (stringBuilder.Length > 0) stringBuilder.Append("-");
			let str = uuidv4();
			if (randomAppend > 0) str = str.substr(0, randomAppend);
			stringBuilder.Append(str);
		}
		switch (ownerType) {
		case OwnerType.Machine:
			stringBuilder.Insert(0, "M-");
			break;
		case OwnerType.User:
			stringBuilder.Insert(0, "U-");
			break;
		case OwnerType.Group:
			stringBuilder.Insert(0, "G-");
			break;
		default:
			throw new Error("Invalid Owner Type");
		}
		return stringBuilder.toString();
	}
}
module.exports = {
	IdUtil,
};
